go/vt/vttablet/tabletserver/state_manager.go

```diff
@@ -392,7 +410,9 @@ func (sm *stateManager) StartRequest(ctx context.Context, target *querypb.Target
 	}
 
 	shuttingDown := sm.wantState != StateServing
-	if shuttingDown && !allowOnShutdown {
+	// If requestsWaitCounter is not zero, then there are go-routines blocked on waiting for requests to be empty.
+	// We cannot allow adding to the requests to prevent any panics from happening.
+	if sm.requestsWaitCounter > 0 || (shuttingDown && !allowOnShutdown) {
```

Comment:
this condition looks wrong to me.
`requestsWaitCounter` will be `0` during shutdown till `request.Wait()` is not called. But we do not want all the request to go through unless `allowOnShutdown` is true.

Reply:
Ignore my comment. I this is good. just that we can reverse the condition check on `or` it may make it more understandable.

Reply:
Sure, I'll reverse them.